File: p1/main.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module Main where

import Prelude hiding (sin)

import Foreign.C -- get the C types
import Foreign.Ptr (Ptr,nullPtr)

import Data.Time.Clock

foreign import ccall safe "runTasks" runTasks :: IO ()
foreign import ccall "runTasks_OpenMP" runTasks_OpenMP :: IO ()
foreign import ccall "initTasks" initTasks :: IO ()

main = do
  initTasks
  profileTask runTasks
  profileTask runTasks_OpenMP
  putStrLn "Done"


profileTask task = do
  startTime <- getCurrentTime
  task
  endTime <- getCurrentTime
  putStrLn (show (diffUTCTime endTime startTime))

File: p1/main.cpp
#include <stdio.h>
#include <time.h>

void initTasks();
void runTasks_OpenMP();
void runTasks();

void profileTask(void (*ptr)());

int main()
{
  initTasks();
  profileTask(runTasks_OpenMP);
  profileTask(runTasks);
  printf("Done\n");
}

void profileTask(void (*ptr)())
{
  clock_t t;
  t = clock();
  (*ptr)();
  t = clock() -t;
  printf("Profile :%f secs\n", ((float)t/CLOCKS_PER_SEC));
}
